hswidget/dccwidget/state.go

```diff
@@ -169,3 +258,52 @@ func (p *widget) makeImagePixel(val byte) color.RGBA {
 
 	return RGBAColor
 }
+
+// nolint:gocognit,gocyclo // will cut later
+func (p *widget) runPlayer(state *widgetState) {
+	for range state.ticker.C {
+		if state.isPlaying {
+			switch state.playMode {
+			case playModeForward:
+				if state.controls.frame < int32(p.dcc.FramesPerDirection-1) {
+					state.controls.frame++
+				} else {
+					if state.repeat {
+						state.controls.frame = 0
+					} else {
+						state.isPlaying = false
+					}
+				}
+			case playModeBackword:
+				if state.controls.frame > 0 {
+					state.controls.frame--
+				} else {
+					if state.repeat {
+						state.controls.frame = int32(p.dcc.FramesPerDirection)
+					} else {
+						state.isPlaying = false
+					}
+				}
+			case playModeLeftRight:
+				if state.leftRightDirection {
+					if fpd := int32(p.dcc.FramesPerDirection) - 1; state.controls.frame < fpd {
+						state.controls.frame++
+						if state.controls.frame == fpd {
+							state.leftRightDirection = false
+						}
+					}
+				} else {
+					if state.controls.frame > 0 {
+						state.controls.frame--
+						if state.controls.frame == 0 {
+							state.leftRightDirection = true
+							if !state.repeat {
+								state.isPlaying = false
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+}
```

Comment:
I acknowledge your `nolint`, but I think 8 layers is too much...
```golang
// wrap integer to max: wrap(450, 360) == 90
func wrap(x, max int) int {
	wrapped := x % max

	if wrapped < 0 {
		return max + wrapped
	}

	return wrapped
}

func (p *widget) runPlayer(state *widgetState) {
	for range state.ticker.C {
		if !state.isPlaying {
			return
		}

		numFrames := int32(p.dcc.FramesPerDirection-1)
		isLastFrame := state.controls.frame == numFrames

		// update play direction
		switch state.playMode {
		case playModeForward:
			state.controls.leftRightDirection = true
		case playModeBackward:
			state.controls.leftRightDirection = false
		case playModeLeftRight:
			if isLastFrame {
				state.controls.leftRightDirection = !state.controls.leftRightDirection
			}
		}

		// now update the frame number
		isForward := state.controls.leftRightDirection == true

		if isForward {
			state.controls.frame++
		} else {
			state.controls.frame--
		}

		state.controls.frame = wrap(state.controls.frame, p.dcc.FramesPerDirection)

		// next, check for stopping/repeat
		isStoppingFrame := (state.controls.frame == 0) || (state.controls.frame == numFrames)

		if isStoppingFrame && !state.repeat {
			state.isPlaying = false
		}
	}
}
```

Reply:
thanx!